### DISCUSSION TITLE
Docker

### DISCUSSION THREAD

**Discussion by Corepex** [coder/coder#4005] [General]
Hi Community, 

I'm using the docker compose docs to setup my coder instance. (https://coder.com/docs/coder-oss/latest/install/docker#run-coder-with-docker-compose)

This works good so far. (web ui is up and running)
BUT how do i init the templates? The docs mention a `coder templates init` command but that doesnt for for me.
Furthermore it would be clever to create a persistent volume for the templates and config stuff ... or? (otherwise every configuration will be lost if you redeploy the container ...)

How do you solve these problems?

**Comment by jsjoeio** [comment]
Hi @Corepex!

Sorry to hear you're having some issues! The initial setup UX definitely needs some love so thanks for sharing this.

> The docs mention a coder templates init command but that doesnt for for me

I just chatted with @bpmct about this and we came up with [some fixes](https://github.com/coder/coder/pull/4027) to make this more clear. Assuming you don't have a domain or externally available, here are the steps I went through:
1. make changes to `docker-compose.yaml` (see [here](https://github.com/coder/coder/pull/4027))
2. run `CODER_TUNNEL=true docker-compose up`
3. look through logs for the `coder login <url>` then run in separate terminal
4. follow prompt and create first user
5. create template with `coder templates init`

Let me know if that helps!

**Reply by bpmct** [reply]
Hey @Corepex - what issue were you running into with `coder templates init`? Was it saying command not found or something else? Even if you have Coder running on Docker, you'll still need to download the CLI locally and log in to Coder via the CLI to manage templates.

re. persistent volumes, you can define [persistent and ephemeral resources](https://coder.com/docs/coder-oss/latest/templates#persistent-vs-ephemeral-resources) in Coder so users files (in certain locations) are persisted, such as in `/home/coder`. For VM-based templates, you can persist the entire VM and just start/stop it.

Here are examples of persistent volumes (`count = 1` or no `count` property):

Docker: 
https://github.com/coder/coder/blob/184e7dbce001ebe11db454079c97a0b798c866ba/examples/templates/docker/main.tf#L64-L66

Kubernetes
https://github.com/coder/coder/blob/184e7dbce001ebe11db454079c97a0b798c866ba/examples/templates/kubernetes/main.tf#L81-L94

**Reply by Corepex** [reply]
Hi @jsjoeio and @bpmct 

thanks for your quick reply. 

according to @jsjoeio information i installed the `coder-cli` locally - so far this works pretty well. 
(added a detailed documentation on how to install the cli on linus [#489](https://github.com/coder/coder-cli/pull/489))

After that i tried to login via `coder login <custom-domain>`.
The command says i should open `https://<custom-domain>/internal-auth?show_token=true` to receive a token that should be entered into the cli. (see screenshot)

If i try to open this page it gives me a "404 This page could not be found." error. 

**Setup:**
Host System (Ubuntu)
|--- coder/coder container (docker)
|------ (image: ghcr.io/coder/coder:latest) 
|
|--- coder/coder database container (docker)
|------ (image: postgres:14.2)

**Note:** 
I installed the coder-cli onto my host system

![image](https://user-images.githubusercontent.com/16717695/190353878-b8a8d4cc-4a27-4e37-ba4e-be4a75322c94.png)

**Reply by jsjoeio** [reply]
Hmm... maybe silly but is your domain externally accessible and if you navigate to the root path, what do you see?

**Reply by bpmct** [reply]
Hey @Corepex. The instructions you've added are for installing the CLI for Coder Classic, our old product. 

The installation process for the CLI is actually the same as for the sever, you can use the install script or a number of different [install methods](https://coder.com/docs/coder-oss/latest/install). From there, instead of starting the server `coder server`, you log into an existing server `coder login <url>`.

**Reply by bpmct** [reply]
With the new CLI, you won't run into a 404 error.

**Reply by Corepex** [reply]
@bpmct thanks the installation of the cli works perfectly with the script. I thought the script would install coder locally - an alternative way to install coder (like system packages or docker).

The initial `coder login <coder-url>` works as it should (and guides me through a "create user" dialog!
The `coder templates init` also works - gives me a punch of predefined options. 
I choose the "docker-code-server" template.

The created parameters file inside the docker-code-server folder matches my system parameters so i run `sudo coder templates create --parameter-file params.sample.yaml`. 

This gives me the following error:
`WebSocket protocol violation: Connection header "close" does not contain Upgrade                                                                                              
{"message":"Failed to accept websocket.","detail":"failed to accept WebSocket connection: WebSocket protocol violation: Connection header \"close\" does not contain Upgrade"}
.                                                                                                                                                                             
Run 'coder templates create --help' for usage. `

Do you know what i did wrong?

**Reply by matifali** [reply]
Hi @Corepex, were you able to do it? This is an old issue. We would be happy to help if you still face this.

**Reply by Corepex** [reply]
> Hi @Corepex, were you able to do it? This is an old issue. We would be happy to help if you still face this.

@matifali Haven't tried anything further. I will let you know if i have any updates 🙌
